### DISCUSSION TITLE
[Worker] Propagate code changes for failed worker to retry

### DISCUSSION THREAD

**Discussion by andreyluchianic** [taskforcesh/bullmq#2107] [Q&A]
I have a fastify server that on request adds Workers that i imported in main.js file as a function - it makes some API requests and stores information into database. 
BullMQ runs on virtual machine with pm2, on a first run (1000 jobs) i've had on virtual machine 2 different version of NodeJS (16 and 18) and because of NodeJS 16 i guess jobs where failing with error: `fetch is not defined` that makes sense cause fetch API exists from NodeJS 18. I've removed old version of Node and updated pm2 Daemon to use version 18. 
Now when i retry jobs some of them are picking up and running OK, but some of them still replying me with same error. I've added debug statements in Worker but changes are NOT propagating for every Worker. For example i've had 300 failed jobs i retried them all smth like 100-120 job went OK with new code and debug info others keep returning same error with same exact stacktrace which should be different if they've run with new code. With each iteration just some part of jobs run with new code. 

How can i force BullMQ to propagate code changes for failed job to retry them ?

<img width="1352" alt="Screenshot 2023-08-02 at 12 25 08" src="https://github.com/taskforcesh/bullmq/assets/88837148/61ec1b2a-eb1c-4328-80d1-e0df1f634fab">

```
// main.mjs
import aiRewriteBullWorker from "./lib/ai_rewrite.mjs";

const openAiWorker = new Worker(queueName, aiRewriteBullWorker, {
  ...redisConnection,
  defaultJobOptions: {
    concurrency: 5,
    attempts: 3,
    backoff: {
      type: 'exponential',
      delay: 1000,
    },
  },
});
```
```
// worker.mjs
const aiRewriteBullWorker = async (job) => {
  ...
};

export default aiRewriteBullWorker;
```

**Comment by manast** [comment]
Not sure what you mean by propagate changes. If you have changed the code of some workers just remove the old workers and start the new ones.

**Reply by andreyluchianic** [reply]
why then some jobs when i retry pickup new code and some doesnt. For example this one 5 or 6 times were failing and them went fine 
<img width="1431" alt="Screenshot 2023-08-02 at 12 47 47" src="https://github.com/taskforcesh/bullmq/assets/88837148/140c40db-ba0a-4734-9f57-b75854a353ed">
it's a bit problematic to recreate new jobs for them cause i'm pulling them out from database in batch

**Reply by manast** [reply]
I don't know, but it seems that it is not related to BullMQ itself but to how you are handling the deployment of your code. Most likely you still have some workers running old code or running on an older version of Node that does not support fetch (which btw it is not a dependency of BullMQ, so that error comes from somewhere else).
